op-supervisor,op-node: Handle non-genesis Interop

sebastianst opened the pull request:
**Description**

The op-node and op-supervisor are taught to handle a non-Genesis Interop activation timestamp properly in managed mode. In essence, the op-supervisor _just does less_ while the op-node does _a bit more_ pre-Interop:

* Before Interop is active, the op-node locally auto-promotes _local_-(un)safe to _cross_(un)safe blocks (as before).
* Only L1 derivation is driven by the op-supervisor pre-Interop.
    * ExhaustL1 events are answered by new L1 blocks as they become available.
    * Reset requests are answered by the supervisor with a special `ResetPreInterop` response.
* When the supervisor detects the (unsafe and safe) activation blocks, it initiates the (events and derivation) DBs.
* If a post-Interop reset bisection cannot find a consistent block, i.e. not even the anchor block stored in the DB is consistent with the node's view, then a pre-Interop reset is initiated as well.

The explicit notion of the anchor block is also mostly removed already in this PR.
* The full supervisor config set includes genesis information, so the op-supervisor tests at startup whether the genesis block already has Interop active. If so, it will locally initiate the DB from genesis information, not querying `AnchorBlock` on the managed nodes.
* If genesis hasn't Interop active yet, it is assumed that the activation block passes by the supervisor via its block subscription on its manages nodes.

This PR also makes the logs DB check itself whether it is initialized so that the ChainsDB initialized markers are only used to denote a fully initialized ChainsDB from a safe activation block.

**Tests**

The action test `TestInteropUpgrade` at `op-e2e/actions/interop/interop_fork_test.go` already runs Interop through an activation and asserts that the Interop activation block is properly sequenced and derived. This implicitly confirms that the activation logic in op-node/op-supervisor functions properly around the Interop activation timestamp. Some modifications were done to this test since it had some false logic that only worked with the previous pseudo-pre-Interop logic that the supervisor was following.

A post-genesis Interop variant of `TestBackendLifetime` was also added, and the previous tests adapted.

More action and acceptance tests will be added in future work.

**Additional context**

In a followup, full reorg support around the activation block is added. It probably wouldn't work right now, will be worked out in https://github.com/ethereum-optimism/optimism/issues/15774

A few proofs tests or checks had to be commented out because the op-challenger (and probably other bits) still need some more work to make non-genesis activation work. This will be resolved with https://github.com/ethereum-optimism/optimism/issues/16166

This PR breaks mutli-node support, which we decided to drop. So a few tests had to be skipped because of this. They will either be removed or fixed (https://github.com/ethereum-optimism/optimism/issues/16187).

This PR also doesn't fully remove the notion of the anchor block yet. This is reserved for future work https://github.com/ethereum-optimism/optimism/issues/16140

**Metadata**

Fixes https://github.com/ethereum-optimism/optimism/issues/13732

sebastianst committed:
op-supervisor,op-node: Handle non-interop genesis activation

protolambda commented on op-node/rollup/interop/managed/api.go:
```diff
@@ -39,6 +39,7 @@ func (ib *InteropAPI) InvalidateBlock(ctx context.Context, seal supervisortypes.
 	return ib.backend.InvalidateBlock(ctx, seal)
 }
 
+// TODO: remove
 func (ib *InteropAPI) AnchorPoint(ctx context.Context) (supervisortypes.DerivedBlockRefPair, error) {
 	return ib.backend.AnchorPoint(ctx)
 }
```
Can this be removed yet?

sebastianst replied:
The anchor point rpc usage is removed in this PR, but we may want to defer full removal to a future PR, to keep the scope and diff of this PR lower. There may still be quite a few references to the anchor point in tests etc, so I think it's worth doing in a follow up instead.

sebastianst replied:
https://github.com/ethereum-optimism/optimism/issues/16140

sebastianst replied:
Added ticket number to TODO

protolambda commented on op-node/rollup/sync/start.go:
```diff
@@ -156,6 +158,8 @@ func FindL2Heads(ctx context.Context, cfg *rollup.Config, l1 L1Chain, l2 L2Chain
 	// then we return the last L2 block of the epoch before that as safe head.
 	// Each loop iteration we traverse a single L2 block, and we check if the L1 origins are consistent.
+	// TODO: maybe check inside here not to go over Interop activation, this should be handled by supervisor
+	// Fine to keep unsafe chain. Safe chain should be capped to stay pre-Interop.
```
This should probably be ticketed

sebastianst replied:
Tracking https://github.com/ethereum-optimism/optimism/issues/16141

protolambda commented on op-supervisor/supervisor/backend/backend.go:
```diff
@@ -176,18 +179,37 @@ func NewSupervisorBackend(ctx context.Context, logger log.Logger,
 		})
 	case superevents.LocalUnsafeUpdateEvent:
+		// TODO: activation block should be directly promoted to cross-safe
+		// because exec msgs aren't allowed in it. Where to best handle this?
```
It should not be cross-safe until it becomes local-safe.
It may however be immediately cross-unsafe after being local-unsafe.

sebastianst replied:
It's a typo, I meant cross-unsafe, and yes I'm aware. My question was "Where to best handle this?" - I think the cleanest way is just to encode this rule in the `cross` package, so that the concern of local-cross promotion is scoped to that package.

sebastianst replied:
This now happens implicitly during DB initialization.

protolambda commented on op-supervisor/supervisor/backend/backend.go:
```diff
@@ -176,18 +179,37 @@ func NewSupervisorBackend(ctx context.Context, logger log.Logger,
+				Safe:    x.Derived,
+			})
+		}
+		// TODO: activation block should be directly promoted to cross-safe
+		// because exec msgs aren't allowed in it. Where to best handle this?
```
Overlaps with comment of above (yes, local-safe to cross-safe is valid here).

sebastianst replied:
I know, my question was "Where to best handle this?" - I think the cleanest way is just to encode this rule in the `cross` package, so that the concern of local-cross promotion is scoped to that package.

sebastianst replied:
This now happens implicitly during DB initialization.

protolambda replied:
Can we clean up the TODO comment then? (and some for above TODO comment also)

sebastianst replied:
done

protolambda commented on op-supervisor/supervisor/backend/db/db.go:
```diff
@@ -163,11 +163,45 @@ func (db *ChainsDB) AttachEmitter(em event.Emitter) {
+		} else {
+			db.logger.Warn("Received unsafe activation block on initialized DB",
+				"chain", x.ChainID, "block", x.Unsafe)
+			// TODO: handle reorg?
```
I think we just need to check whether or not the regular reorg-detection runs at all, since there's a `UnsafeActivationBlockEvent` here, instead of a regular block event.

sebastianst replied:
I want to handle full activation block reorg support in a follow up, tracking with https://github.com/ethereum-optimism/optimism/issues/15774 (existing ticket, just reworded the title).

protolambda commented on op-supervisor/supervisor/backend/db/db.go:
```diff
@@ -163,11 +163,45 @@ func (db *ChainsDB) AttachEmitter(em event.Emitter) {
 	case superevents.LocalDerivedEvent:
+		if !db.isInitialized(x.ChainID) {
+			// This case is handled by SafeActivationBlockReceivedEvent
+			// TODO: Critical error, we should never receive this event before the DB is initialized
+			//  Signal reset/safe-data-needed via event?
```
Some reset trigger may work. At least we should log what happens here, so we know what went wrong, if expectations are ever broken.

sebastianst replied:
I realize that the activation block will also land here, so this branch is expected. I will just debug log for now. We don't have the full chain cfg available in the ChainsDB, so can't check here if it's an activation block for better logging... Other checks will trigger a reset anyways.

protolambda commented on op-supervisor/supervisor/backend/db/update.go:
```diff
@@ -226,11 +229,16 @@ func (db *ChainsDB) onFinalizedL1(finalized eth.BlockRef) {
 	db.logger.Info("Updated finalized L1", "finalizedL1", finalized)
 	db.finalizedL1.Unlock()
 
+	// TODO: There seems to be no consumer of this event?
 	db.emitter.Emit(superevents.FinalizedL1UpdateEvent{
```
I think the `status` package may want to listen for this event.

sebastianst replied:
Doesn't look like, the `FinalizedL1UpdateEvent` type isn't referenced anywhere else but here, in particular also not in `status`. Is it missing and should be handled there?

protolambda replied:
I think adding a case to the switch in `status`, and logging it, would be nice.
And demote the log level here in the db code to debug level.

Generally I hope we can make all info-logs more deliberate, in one place. So then it is easy to adjust noisy-ness. And just have the packages do debug logs.
For finality updates it's not a big deal, but for other more frequent updates it is nice.

sebastianst replied:
OK now receiving this event in the `StatusTracker` but only Info-logging it there (using global logger, it doesn't have a logger attached). Downgraded log message in ChainsDB to debug. We could also add the finalized L1 to the `SupervisorSyncStatus` as a new field. But would defer this to a follow up.

protolambda commented on op-supervisor/supervisor/backend/syncnode/rpc.go:
```diff
@@ -151,13 +154,24 @@ func (rs *RPCSyncNode) Reset(ctx context.Context, lUnsafe, xUnsafe, lSafe, xSafe
 }
 
 func (rs *RPCSyncNode) AnchorPoint(ctx context.Context) (types.DerivedBlockRefPair, error) {
-	var out types.DerivedBlockRefPair
+	var (
```
Are we still using the anchor point RPC?

sebastianst replied:
The anchor point rpc usage is removed in this PR, but we may want to defer full removal to a future PR, to keep the scope and diff of this PR lower. There may still be quite a few references to the anchor point in tests etc, so I think it's worth doing in a follow up instead.

sebastianst replied:
Tracking with https://github.com/ethereum-optimism/optimism/issues/16140

protolambda commented on op-supervisor/supervisor/backend/db/anchor.go:
```diff
@@ -75,21 +75,25 @@ func (db *ChainsDB) maybeInitSafeDB(id eth.ChainID, anchor types.DerivedBlockRef
 	} else {
 		logger.Debug("Events database already initialized")
-		if seal.Hash != anchor.Derived.Hash {
+		// TODO: to allow for reorgs, we may want to handle this differently
+		if seal.Hash != anchor.Hash {
```
If the rewinder can detect the difference between op-node chain and supervisor chain, and rewind out the initial starting block in the supervisor db, then we don't need to reorg here.

This is captured by the rewind activation block github issue though.
Maybe update the TODO comment reference with issue number of https://github.com/ethereum-optimism/optimism/issues/15774

sebastianst replied:
updated comment and added issue number

protolambda commented on op-supervisor/supervisor/backend/db/events.go:
```diff
@@ -0,0 +1 @@
+package db
```
What is this empty file for? Can we delete it?

sebastianst replied:
yes, leftover from a previous experiment

protolambda commented on op-supervisor/supervisor/backend/processors/chain_processor.go:
```diff
@@ -94,6 +96,7 @@ func (s *ChainProcessor) AddSource(cl Source) {
 func (s *ChainProcessor) nextNum() uint64 {
 	headNum, ok := s.rewinder.LatestBlockNum(s.chain)
 	if !ok {
+		// TODO: need to check what we should do here for non-genesis interop
```
I think we should abort indexing from 0. We could wait for the activation-block event, and use that to set a starting-point, if rewinder doesn't have a starting-block for us.

sebastianst replied:
I'm now checking in `onRequest` if we returned a 0 here, which always decodes for an empty DB, and then don't start processing, only warn-log. This would be very unexpected anyways, as the process event is not emitted for the activation block. The activation block will eventually arrive and then processing should continue to work as usual. This is better than processing from `0`, which is a block number we never want to process, even in the genesis case.

protolambda commented on op-supervisor/supervisor/backend/db/logs/db.go:
```diff
@@ -193,8 +199,10 @@ func (db *DB) OpenBlock(blockNum uint64) (ref eth.BlockRef, logCount uint32, exe
 	defer db.rwLock.RUnlock()
 
+	// TODO: need to fix assumption that the first block is block 0
+	// This is only true for genesis-Interop chains.
 	if blockNum == 0 {
-		seal, err := db.StartingBlock()
+		seal, err := db.FirstSealedBlock()
```
`newIteratorAt` below handles the starting block case for non-genesis case. This `== 0` check is just to prevent the underflow.

sebastianst replied:
That makes sense. But shouldn't we still change below logic to return an error `if seal.Number != 0`? Right now we just warn-log and return the block with a number > 0, which is not a defined behavior of `OpenBlock`. If this behavior is wanted from `OpenBlock`, we should document it. Otherwise let's return an error here. Maybe even a new error type like `ErrPreInterop` or `ErrPreActivation` to denote situations like this (the opposite of `ErrFuture`).

protolambda replied:
Fixed. And using the `types.ErrSkipped` error, which is meant for the kind of access to older data that is not indexed.

sebastianst commented on op-supervisor/supervisor/backend/cross/safe_update.go:
```diff
@@ -73,7 +73,13 @@ func CrossSafeUpdate(logger log.Logger, chainID eth.ChainID, d CrossSafeDeps, li
 	}
-	crossSafeRef := currentCrossSafe.Derived.MustWithParent(parent.ID())
+	var crossSafeRef eth.BlockRef
+	// During non-genesis Interop activation, the parent may be zero, which is ok.
+	if parent.ID() == (eth.BlockID{}) {
+		crossSafeRef = currentCrossSafe.Derived.WithZeroParent()
+	} else {
+		crossSafeRef = currentCrossSafe.Derived.MustWithParent(parent.ID())
+	}
```
I thought about just using `ForceWithParent` for brevity, but then we wouldn't have the proper parent checks for non-zero parents.

protolambda committed:
op-supervisor: fix OpenBlock error case

sebastianst committed:
op-e2e: Skip tests that depend on multi-node setup

Multi-node will be completely removed via #16174.
Other tests will be fixed by switching to multi-supervisor setup (#16187).

ajsutton commented on op-e2e/actions/interop/proofs_test.go:
```diff
@@ -176,6 +176,8 @@ func TestInteropFaultProofs_ConsolidateValidCrossChainMessage(gt *testing.T) {
 
 func TestInteropFaultProofs_PreForkActivation(gt *testing.T) {
+	// TODO(#16166): Fix non-genesis Interop activation proofs
+	gt.Skip()
```
For the subtests that are failing:
`FirstChainOptimisticBlock-challenger` - op-supervisor is incorrectly returning not found from `optimism_superRootAtTimestamp` (because it gets `cross-derived-to-source failed for chain 900200: future data`).

The other three are because challenger is now expecting that the prestate is the "invalid" claim which is used when supervisor [returns not found for the agreed prestate](https://github.com/ethereum-optimism/optimism/blob/4a27936ce8293d09ecb610ef5d243988301f2f78/op-challenger/game/fault/trace/super/provider.go#L96-L99). The difference between these and the first one is that in the first case its getting the absolute prestate so it should always be valid and if it gets not found it returns that as an error, whereas later claims might actually be for blocks past the current unsafe head and thus legitimately return not found which should then be translated to `InvalidTransition`.

ajsutton replied:
So the error is coming from [the call into `CrossDerivedToSource`](https://github.com/ethereum-optimism/optimism/blob/5c609beab71385ef183d1ab0527ec72009c91a4f/op-supervisor/supervisor/backend/backend.go#L747) as part of creating the super root which is used to provide information on which L1 block the super root was first safe at.  That returns `types.ErrSkipped` now, I'm guessing because the PR actually has the supervisor skipping all data prior to interop activating.

The problem is that we _do_ need to provide the data on when things became safe so the challenger can respond correctly to games where unsafe blocks are proposed (ie the super root is valid but the game was created before the batch data was submitted to L1). Those proposals are invalid.  The `CrossSafeDerivedFrom` field in the super root response is important for this as is the `AllSafeDerivedAt` function returning the correct local safe head that would be derived for a given L1 hash. My guess is that would be broken by these changes as well.

Prior to interop that data came from the safe head db in op-node, so we could potentially defer to that when the lookup returns `ErrSkipped`.

sebastianst replied:
https://github.com/ethereum-optimism/optimism/issues/16166

ajsutton commented on op-e2e/actions/interop/interop_fork_test.go:
```diff
@@ -115,14 +110,21 @@ func TestInteropUpgrade(gt *testing.T) {
 
 	// Verify proofs agree
-	assertProgramOutputMatchesDerivationForBlockTimestamp(gt, system, system.Actors.ChainA.Sequencer.L2Unsafe().Time)
+	// TODO(#16166): Fix non-genesis Interop activation proofs
+	// assertProgramOutputMatchesDerivationForBlockTimestamp(gt, system, system.Actors.ChainA.Sequencer.L2Safe().Time)
```
This is actually the same failure case as above, supervisor is failing to return a super root that we're requesting:

```
        	            	cross-derived-to-source failed for chain 900200: query is before first entry LinkEntry(source: BlockSeal(hash:0xd7073450474d7087067af9543164d2bf421a088bac8dc70e5940101ee7f74511, number:3, time:1748568053), derived: BlockSeal(hash:0xb2ac632babb01bf236db4e88455268acfb51921a0e8baf800ddb0b91ec7774b0, number:2, time:1748568021), invalidated: false): skipped data
```

sebastianst replied:
Will be resolved with https://github.com/ethereum-optimism/optimism/issues/16166

ajsutton reviewed (commented):
Took a look at the challenger failures and it looks like this change has broken the key requirement challenger has on it's API that `optimism_superRootAtTimestamp` will return a valid super root for any timestamp up to at least the current safe head.  It is critical that it can return a super root even prior to interop activating, because the new anchor state we set will be from before interop hard fork activates and any block after that can be disputed. So challenger has to be able to get the valid super root for timestamps from that anchor state forwards in order to post correct claims.
